File: agent-runner/src/mcp/tools/qa.ts
// agent-runner/src/mcp/tools/qa.ts
import { tool } from "@anthropic-ai/claude-agent-sdk";
import { z } from "zod";
import type { WorkspaceContext } from "../context.js";
import { fetchWithRetry, formatFetchError, buildHeaders, BACKEND_URL } from "./fetch-utils.js";

export function createQATools(context: WorkspaceContext) {
  return [
    tool(
      "request_user_browser_action",
      "Request the user to perform an action in a browser that requires human interaction (e.g., logging in, completing OAuth flow, solving a CAPTCHA). " +
      "The agent will pause until the user completes the action and clicks Done. " +
      "Use this when automated testing hits an authentication wall or other human-required interaction.",
      {
        url: z.string().describe("The URL the user should navigate to"),
        instructions: z.string().describe("Clear instructions for what the user needs to do (e.g., 'Log in with your credentials and navigate to the dashboard')"),
        testCase: z.string().optional().describe("The test case context — what you're trying to test after the user completes this action"),
      },
      async ({ url, instructions, testCase }) => {
        try {
          // The agent-runner will receive this as a pending QA handoff via the streaming snapshot.
          // The frontend will display a QAHandoffPrompt to the user.
          // When the user responds, the backend sends the response back through the process channel.
          const response = await fetchWithRetry(
            `${BACKEND_URL}/api/repos/${context.workspaceId}/sessions/${context.sessionId}/qa-handoff`,
            {
              method: "POST",
              headers: buildHeaders(true),
              body: JSON.stringify({ url, instructions, testCase }),
            }
          );

          if (!response.ok) {
            // If the endpoint doesn't exist yet, return a helpful message
            return {
              content: [{
                type: "text" as const,
                text: `QA handoff requested. Please ask the user to:\n\n1. Open: ${url}\n2. ${instructions}\n\nOnce done, they can continue the conversation.\n\n${testCase ? `Test context: ${testCase}` : ''}`,
              }],
            };
          }

          const result = await response.json();
          if (result.completed) {
            return {
              content: [{
                type: "text" as const,
                text: `User completed the browser action.${result.notes ? `\n\nUser notes: ${result.notes}` : ''}\n\nYou can now continue with: ${testCase || 'the test'}`,
              }],
            };
          } else {
            return {
              content: [{
                type: "text" as const,
                text: `User skipped the browser action.${result.notes ? `\n\nUser notes: ${result.notes}` : ''}\n\nYou may need to adjust your test approach.`,
              }],
            };
          }
        } catch (error) {
          return {
            content: [{
              type: "text" as const,
              text: `Error requesting QA handoff: ${formatFetchError(error)}.\n\nFallback: Please ask the user directly to open ${url} and ${instructions}`,
            }],
          };
        }
      },
      { annotations: { readOnlyHint: false, destructiveHint: false, openWorldHint: true } }
    ),
  ];
}

File: backend/loop/chatml/qa.go
package chatml

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/chatml/chatml-core/tool"
)

// --- request_user_browser_action ---

type requestUserBrowserActionTool struct {
	ctx *ToolContext
}

func (t *requestUserBrowserActionTool) Name() string           { return "mcp__chatml__request_user_browser_action" }
func (t *requestUserBrowserActionTool) IsConcurrentSafe() bool { return true }
func (t *requestUserBrowserActionTool) DeferLoading() bool     { return true }
func (t *requestUserBrowserActionTool) Description() string {
	return "Request the user to perform an action in a browser that requires human interaction (e.g., login, OAuth, CAPTCHA)."
}
func (t *requestUserBrowserActionTool) InputSchema() json.RawMessage {
	return json.RawMessage(`{
		"type": "object",
		"properties": {
			"url": {"type": "string", "description": "The URL the user should navigate to"},
			"instructions": {"type": "string", "description": "Clear instructions for what the user needs to do"},
			"testCase": {"type": "string", "description": "The test case context"}
		},
		"required": ["url", "instructions"]
	}`)
}

func (t *requestUserBrowserActionTool) Execute(ctx context.Context, input json.RawMessage) (*tool.Result, error) {
	var params struct {
		URL          string `json:"url"`
		Instructions string `json:"instructions"`
		TestCase     string `json:"testCase"`
	}
	if err := json.Unmarshal(input, &params); err != nil {
		return tool.ErrorResult(fmt.Sprintf("invalid input: %v", err)), nil
	}

	// V1: Return instruction text for the agent to present to the user.
	// Full implementation with WebSocket pause/resume is deferred.
	var content string
	if params.TestCase != "" {
		content = fmt.Sprintf("Please perform the following browser action:\n\nURL: %s\nInstructions: %s\nTest case: %s\n\nPlease complete this action and let me know when you're done.", params.URL, params.Instructions, params.TestCase)
	} else {
		content = fmt.Sprintf("Please perform the following browser action:\n\nURL: %s\nInstructions: %s\n\nPlease complete this action and let me know when you're done.", params.URL, params.Instructions)
	}

	return &tool.Result{Content: content}, nil
}
